pkg/v1/remote/transport/bearer.go

```diff
@@ -230,20 +229,29 @@ func (bt *bearerTransport) Refresh(ctx context.Context, auth *authn.AuthConfig)
 		content []byte
 		err     error
 	)
+
+	first, second := bt.refreshBasic, bt.refreshOauth
```

Comment:
Let me know if I understand the intention of this change correctly:

Previously, we would only enter `refreshOauth` if `IdentityToken` was set, which means we would ignore `Username` and `Password` (hence, the TODO I left).

Now, (if `IdentityToken` is unset) we will try `refreshBasic` first, falling back to `refreshOauth` if we encounter a 404 or 405, so we will end up going down the Username/Password path after `refreshBasic` fails.

If `IdentityToken` _is_ set, we reverse that order and first try `refreshOauth`, falling back to `refreshBasic`, which matches the current behavior.

Can you provide any evidence that this is the right thing to do on 404/405? (I believe you, I would just love to have a comment in here somewhere as justification, even if it's just something simple like "foo.example.com returns a 404/5 in this case").

Reply:
This logic somewhat copies what `containerd` does: https://github.com/containerd/containerd/blob/v1.7.2/remotes/docker/authorizer.go#L297-L321
`containerd` is a commonly used container runtime in kubernetes environments and it's well battle tested, so I believe it to be a good source of truth in the absence of an RFC that specifies when to use oauth vs basic.

The only difference is the preference for the basic flow when an username/password is provided and for the oauth flow when an identity token is provided. `containerd` always tries oauth first. I made it this way because it seems this repo had a prior art that did this: https://github.com/google/go-containerregistry/pull/566

I personally think we should copy what `containerd` does (i.e. always try oauth first) and I'm happy to change, but I'm also happy to hear why did we choose to select the flow preference based on the type of credentials provided.

Reply:
> I personally think we should copy what containerd does (i.e. always try oauth first) and I'm happy to change, but I'm also happy to hear why did we choose to select the flow based on the type of credentials provided.

I'd like to keep basic first, mostly because it's more widely supported and will reduce the number of roundtrips for registries that don't support the oauth flow. This is at the expense of registries that _only_ support the oauth flow, but that doesn't affect me, personally, so I'd selfishly bias towards basic first 😅

Reply:
That sounds ok to me. The registry I'm working with should behave nicely by returning a 405 if the method is a GET on the token endpoint.

I believe the implementation suffices as is, or would you like to use basic first if we see an identity token as well?
Also, are there any other changes you'd like to see in the PR?